DOL.WHD.Section14c.Api/Providers/ApplicationOAuthProvider.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using DOL.WHD.Section14c.DataAccess.Identity;
+using DOL.WHD.Section14c.Domain.Models;
+using Microsoft.AspNet.Identity.Owin;
+using Microsoft.Owin.Security;
+using Microsoft.Owin.Security.Cookies;
+using Microsoft.Owin.Security.OAuth;
+
+namespace DOL.WHD.Section14c.Api.Providers
+{
+    public class ApplicationOAuthProvider : OAuthAuthorizationServerProvider
+    {
+        private readonly string _publicClientId;
+
+        public ApplicationOAuthProvider(string publicClientId)
+        {
+            if (publicClientId == null)
+            {
+                throw new ArgumentNullException("publicClientId");
+            }
+
+            _publicClientId = publicClientId;
+        }
+
+        public override async Task GrantResourceOwnerCredentials(OAuthGrantResourceOwnerCredentialsContext context)
+        {
+            var userManager = context.OwinContext.GetUserManager<ApplicationUserManager>();
+
+            ApplicationUser user = await userManager.FindByNameAsync(context.UserName);
+            if (user != null)
+            {
+                var passwordExpired = false;
+                var passwordExpirationDays = Convert.ToInt32(ConfigurationManager.AppSettings["PasswordExpirationDays"]);
+                if (passwordExpirationDays > 0)
+                {
+                    passwordExpired = user.LastPasswordChangedDate.AddDays(passwordExpirationDays) < DateTime.Now;
+                } 
+                var validCredentials = await userManager.FindAsync(context.UserName, context.Password);
+                if (await userManager.IsLockedOutAsync(user.Id))
+                {
+                    // account locked
+                    // use invalid user name or password message to avoid disclosing that a valid username was input
+                    context.SetError("invalid_grant", "The user name or password is incorrect.");
+                }
+                else if (validCredentials == null)
+                {
+                    // invalid credentials
+                    // increment failed login count
+                    if (await userManager.GetLockoutEnabledAsync(user.Id))
+                    {
+                        await userManager.AccessFailedAsync(user.Id);
+                    }
+
+                    context.SetError("invalid_grant", "The user name or password is incorrect.");
```

Comment:
Moved to resource file. https://github.com/18F/dol-whd-14c/pull/67/commits/737c964e525fae15e5c3d4bca07c16c6d52d0f1e